[FAB-17266] Move CouchDB tutorial to test network (#473)
hyperledger/fabric: docs/source/couchdb_tutorial.rst

topic. Follow the tutorial below for details on how to leverage CouchDB in your
blockchain network.

Throughout this tutorial, we will use the `Marbles sample <https://github.com/hyperledger/fabric-samples/blob/master/chaincode/marbles02/go/marbles_chaincode.go>`__
as our use case to demonstrate how to use CouchDB with Fabric and will deploy
Marbles to the Fabric test network. You should have completed the task
:doc:`install`.

Why CouchDB?
~~~~~~~~~~~~

Fabric supports two types of peer databases. LevelDB is the default state
database embedded in the peer node. LevelDB stores chaincode data as simple
key-value pairs and only supports key, key range, and composite key queries.

> CouchDB is an optional alternate state database that allows you to model data

Reviewer: ```suggestion
CouchDB is an optional, alternate state database that allows you to model data
```

Author: done.

on the ledger as JSON and issue rich queries against data values rather than
the keys. CouchDB also allows you to deploy indexes with your chaincode to make

> queries more efficient, and enable you to query large datasets.

Reviewer: ```suggestion
queries more efficient and enable you to query large datasets.
```

Author: done.

> CouchDB runs as a separate database process alongside the peer, therefore there

Reviewer 2: ```suggestion
CouchDB runs as a separate database process alongside the peer, there
```

Author: done.

are additional considerations in terms of setup, management, and operations.

> A Docker image of `CouchDB <https://hub.docker.com/r/hyperledger/fabric-couchdb/>`__

Reviewer 2: ```suggestion
A Docker image of `CouchDB <https://hub.docker.com/_/couchdb/>`__
```
We are no longer building or shipping our own CouchDB image.

Author: done.

Edit the ``stateDatabase`` section of ``core.yaml``. Specify ``CouchDB`` as the
``stateDatabase`` and fill in the associated ``couchDBConfig`` properties. For

> more information, see `CouchDB configuration <http://hyperledger-fabric.readthedocs.io/en/master/couchdb_as_state_database.html#couchdb-configuration>`__.

Reviewer 2: Do we need to hard-code the link to "master"?

Author: switching to relative links.

Add the index to your chaincode folder
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Once you finalize an index, you need to package it with your chaincode for

> deployment by being placed alongside it in the appropriate metadata folder. You

Reviewer: ```suggestion
deployment by placing it in the appropriate metadata folder. You
```

This also seems to overlap with the explicit directions that were added below. Seems we would only need one.

Author: done.

:guilabel:`Try it yourself`

We will bring up the Fabric test network and use it to deploy the marbles
chainocde. Use the following command to navigate to the `test-network` directory
in the Fabric samples:

.. code:: bash

    cd fabric-samples/test-network

For this tutorial, we want to operate from a known initial state. The following
command will kill any active or stale Docker containers and remove previously
generated artifacts:

.. code:: bash

    ./network.sh down

If you have not run through the tutorial before, you will need to vendor the
chaincode dependencies before we can deploy it to the network. Run the
following commands:

.. code:: bash

    cd ../chaincode/marbles02/go
    GO111MODULE=on go mod vendor
    cd ../../../test-network

From the `test-network` directory, deploy the test network with CouchDB with the
following command:

.. code:: bash

    ./network.sh up createChannel -c mychannel -s couchdb

> 

Reviewer 3: Since you default to channel name mychannel there is really no reason to include that here. We don't include it in the private data tutorial, so in the interest of consistency we should maybe just remove it here?

Author: done

endorse our transactions. However, before we can interact with our chaincode,
the members of the channel need to agree on a chaincode definition that
establishes chaincode governance. In the previous section, we demonstrated how
to add the index to the chaincode folder so that the index is deployed with

> the chaincode .

Reviewer: ```suggestion
the chaincode.
```

Author: done.

The chaincode needs to be packaged before it can be installed on our peers.
We can use the `peer lifecycle chaincode package <http://hyperledger-fabric.readthedocs.io/en/latest/commands/peerlifecycle.html#peer-lifecycle-chaincode-package>`__ command
to package the marbles chaincode.

:guilabel:`Try it yourself`

1. After you start the test network, copy and paste the following environment
variables in your CLI to interact with the network as the Org1 admin. Make sure
that you are in the `test-network` directory.

.. code:: bash

    export PATH=${PWD}/../bin:${PWD}:$PATH
    export FABRIC_CFG_PATH=$PWD/../config/
    export CORE_PEER_TLS_ENABLED=true

>     export CORE_PEER_LOCALMSPID="Org1MSP"

Reviewer: nit: double quotes should be used around the `$PWD` references. I wouldn't bother changing it here but may be worth a separate task for some other time.

Author: done.

JSON file and add or remove fields from the index. Fabric only supports the

> index type JSON, changing the index type is not supported. The updated

Reviewer 2: ```suggestion
index type JSON.  Changing the index type is not supported. The updated
```

Author: Will do

> .. note:: The Fauxton interface is a web UI for the creation, update, and

Reviewer 2: I know this was here before, but do we really need to have explicit doc on how to use CouchDB tools?

Author: From past experience, Fabric users are not usually experienced database admins and may need a push in the right direction. See the best practices for queries and indexes section above, which was put in based on user feedback. It also makes the tutorial a bit more fun. You can go to your browser and see your indexes. The note is far down enough that it does not clutter the tutorial.

Reviewer 2: ok
